### PR TITLE
Fixes #2450

### LINKED ISSUE #2450: Pre-condition Violation in QueryAtom.GetSmarts()
**Description:**

- RDKit Version: 2019.03.1
- Platform: Mac OS

The GetSmarts method of a QueryAtom object raises a "Pre-condition Violation":

```
>>> from rdkit.Chem import rdqueries
>>> qatom = rdqueries.AtomNumEqualsQueryAtom(6)
>>> qatom.GetSymbol()
'*'
>>> qatom.GetSmarts()
[10:50:05] 

****
Pre-condition Violation
no owner
Violation occurred on line 124 in file /Users/glandrum/anaconda5/conda-bld/rdkit_1554778814548/work/Code/GraphMol/Atom.h
Failed Expression: dp_mol
****

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Pre-condition Violation
        no owner
        Violation occurred on line 124 in file Code/GraphMol/Atom.h
        Failed Expression: dp_mol
        RDKIT: 2019.03.1
        BOOST: 1_67

```

This piece of code did work with a previous version of RDKit (although I did not remember exactly which one).

As a workaround, it is possible to extract an "already created" QueryAtom from a RDKIt Mol created : 

```
>>> from rdkit.Chem import MolFromSmarts
>>> m = MolFromSmarts('[*]')
>>> qatom = m.GetAtomWithIdx(0)
>>> qatom.GetSymbol()
'*'
>>> qatom.GetSmarts()
'*'
```

Thanks!

### PR DESCRIPTION
**Pull request by greglandrum**
Adds `hasOwningMol()` to a few objects.

Note that `SmilesWrite::GetAtomSmiles()` still doesn't work (and won't) since it relies upon information about valence that's only available for atoms associated with molecules. We could think about changing that, but that's a bigger project.

### CODE REVIEW

File: Code/GraphMol/Conformer.h [hunk 1]
```diff
@@ -71,7 +70,10 @@ class RDKIT_GRAPHMOL_EXPORT Conformer : public RDProps {
   //! Reserve more space for atom position
   void reserve(unsigned int size) { d_positions.reserve(size); }
 
-  //! Get the molecule that oqns this conformation
+  //! returns whether or not this instance belongs to a molecule
+  bool hasOwningMol() const { return dp_mol != nullptr; };
+
+  //! Get the molecule that owns this instance
   ROMol &getOwningMol() const { return *dp_mol; }
```
Comment (bp-kelley) [inline Code/GraphMol/Conformer.h]: Needs PRECONDITION now that we have hasOwningMol, i.e. this implies that we may not have an owning mol :)

**Comment by bp-kelley** [comment]
For atomGetSmiles I'm assuming this is a plumbing issue with hydrogens and such.

**Comment by greglandrum** [comment]
> For atomGetSmiles I'm assuming this is a plumbing issue with hydrogens and such.

Yessir. That's exactly it.

File: Code/GraphMol/Conformer.h (hunk 1 above)
Comment (greglandrum) [reply Code/GraphMol/Conformer.h]: yikes. You are completely correct.

Comment (greglandrum) [reply Code/GraphMol/Conformer.h]: Ok, add the `PRECONDITION` here and in `SubstanceGroup`

**Pull request merged**